zebra-chain/src/transaction/sighash.rs

```diff
@@ -0,0 +1,924 @@
+effectively means that all transaction signatures from the Sapling activation
+height (as specified in 6) will use the new algorithm.";
+
+const OVERWINTER_VERSION_GROUP_ID: u32 = 0x03C4_8270;
+const SAPLING_VERSION_GROUP_ID: u32 = 0x892F_2085;
+
+const ZCASH_SIGHASH_PERSONALIZATION_PREFIX: &[u8; 12] = b"ZcashSigHash";
+const ZCASH_PREVOUTS_HASH_PERSONALIZATION: &[u8; 16] = b"ZcashPrevoutHash";
+const ZCASH_SEQUENCE_HASH_PERSONALIZATION: &[u8; 16] = b"ZcashSequencHash";
+const ZCASH_OUTPUTS_HASH_PERSONALIZATION: &[u8; 16] = b"ZcashOutputsHash";
+const ZCASH_JOINSPLITS_HASH_PERSONALIZATION: &[u8; 16] = b"ZcashJSplitsHash";
+const ZCASH_SHIELDED_SPENDS_HASH_PERSONALIZATION: &[u8; 16] = b"ZcashSSpendsHash";
+const ZCASH_SHIELDED_OUTPUTS_HASH_PERSONALIZATION: &[u8; 16] = b"ZcashSOutputHash";
+
+bitflags::bitflags! {
+    pub struct HashType: u32 {
+        const ALL = 0b0000_0001;
+        const NONE = 0b0000_0010;
+        const SINGLE = Self::ALL.bits | Self::NONE.bits;
+        const ANYONECANPAY = 0b1000_0000;
+    }
+}
+
+impl HashType {
+    fn masked(self) -> Self {
+        Self::from_bits_truncate(self.bits & 0b0001_1111)
+    }
+}
+
+pub(super) struct SigHasher<'a> {
+    trans: &'a Transaction,
+    hash_type: HashType,
+    network_upgrade: NetworkUpgrade,
+    input: Option<(transparent::Output, &'a transparent::Input, usize)>,
+}
+
+impl<'a> SigHasher<'a> {
+    pub fn new(
+        trans: &'a Transaction,
+        hash_type: HashType,
+        network_upgrade: NetworkUpgrade,
+        input: Option<(u32, transparent::Output)>,
+    ) -> Self {
+        let input = if let Some((index, prevout)) = input {
+            let index = index as usize;
+            let inputs = trans.inputs();
+
+            Some((prevout, &inputs[index], index))
+        } else {
+            None
+        };
+
+        SigHasher {
+            trans,
+            hash_type,
+            network_upgrade,
+            input,
+        }
+    }
+
+    pub fn sighash(self) -> Hash {
+        use NetworkUpgrade::*;
+        let mut hash = blake2b_simd::Params::new()
+            .hash_length(32)
+            .personal(&self.personal())
+            .to_state();
+
+        match self.network_upgrade {
+            Genesis | BeforeOverwinter => unreachable!("Zebra checkpoints on Sapling activation"),
+            Overwinter => self
+                .hash_sighash_zip143(&mut hash)
+                .expect("serialization into hasher never fails"),
+            Sapling | Blossom | Heartwood | Canopy => self
+                .hash_sighash_zip243(&mut hash)
+                .expect("serialization into hasher never fails"),
+        }
+
+        hash.finalize()
+    }
+
+    fn consensus_branch_id(&self) -> ConsensusBranchId {
+        self.network_upgrade
+            .branch_id()
+            .expect("Zebra checkpoints on Sapling activation")
+    }
+
+    /// Sighash implementation for the overwinter consensus branch
```

Comment:
Nit: network upgrade is the more common term

```suggestion
    /// Sighash implementation for the overwinter network upgrade
```